### ISSUE TITLE
declarative einsum initialization with walrus operator

### ISSUE THREAD

**Issue by bionicles** [numpy/numpy#18590]
## Feature
 I like the clarity of the code from [Julia's Tullio.jl](https://github.com/mcabbott/Tullio.jl). Unsure if it could be done in Numpy. I just find numpy's einsum notation to be a bit hard to follow 

Is it theoretically possible to implement something like this with Numpy?
```jl
@tullio M[x,y,c] := N[x+i, y+j,c] * K[i,j]     # sum over i,j, and create M

@tullio S[x] = P[x,y] * log(Q[x,y] / R[y])     # sum over y, and write into S

@tullio A[i,j] += B[i,k,l] * C[l,j] * D[k,j]   # sum over k,l, and add to values in A

@tullio (*) Z[j] := X[ind[k],j] * exp(-Y[k])   # product over k
```

**Comment by seberg** [comment]
Yeah, I have to agree.  Sorry, but while it is cute syntax, I don't think its a concrete enough proposal for the issue tracker.  I don't know if there is some python magic possible (but I doubt it at this point).  That is because it seems like this would have to dig down and modify the AST, since you are proposing to use the `[]` syntax.

As Eric said, a string-syntax is possible, but at that point there is no requirement to start doing it inside NumPy proper, so the easier route would be to experiment outside and move it in if it turns out very successful.
